base/inference.jl

```diff
@@ -821,7 +810,109 @@ function abstract_apply(af, fargs, aargtypes::Vector{Any}, vtypes, sv, e)
     return abstract_call(af, (), Any[Vararg{Any}], vtypes, sv, ())
 end
 
+function isconstantargs(args, argtypes::Vector{Any}, sv::StaticVarInfo)
+    if isempty(argtypes)
+        return true
+    end
+    if is(args,()) || isa(argtypes[end], Vararg)
+        return false
+    end
+    for i = 1:length(args)
+        arg = args[i]
+        t = argtypes[i]
+        if !isType(t) || has_typevars(t.parameters[1])
+            if isconstantref(arg, sv) === false
+                return false
+            end
+        end
+    end
+    return true
+end
+
+function _ieval_args(args, argtypes::Vector{Any}, sv::StaticVarInfo)
+    c = cell(length(args))
+    for i = 1:length(argtypes)
+        t = argtypes[i]
+        if isType(t) && !has_typevars(t.parameters[1])
+            c[i] = t.parameters[1]
+        else
+            c[i] = _ieval(isconstantref(args[i], sv))
+        end
+    end
+    return c
+end
+
+@pure function type_typeof(v::ANY)
+    if isa(v, Type)
+        return Type{v}
+    end
+    return typeof(v)
+end
+
+function pure_eval_call(f, fargs, argtypes, sv, e)
+    if !isa(f, Function) # TODO: maybe replace with :call?
+        return false
+    end
+    if !isconstantargs(fargs, argtypes, sv)
+        return false
+    end
+
+    args = _ieval_args(fargs, argtypes, sv)
+    tm = _topmod()
+    if isgeneric(f)
+        atype = Tuple{Any[type_typeof(a) for a in args]...}
+        meth = _methods(f, atype, 1)
+        if meth === false || length(meth) != 1
+            return false
+        end
+        meth = meth[1]::SimpleVector
+        linfo = try
+            func_for_method(meth[3], atype, meth[2])
+        catch
+            NF
+        end
+        if linfo === NF
+            return false
+        end
+        if !linfo.pure
+            typeinf(linfo, meth[1], meth[2], linfo)
+            if !linfo.pure
+                return false
+            end
+        end
+    elseif !isdefined(f, :code) || !f.code.pure
+        return false
+    end
+
+    local v
+    try
+        v = f(args...)
```

Comment:
@carnval i'm not convinced that those require different annotations. it's true that's how it is annotated today, but I think there are a number of heuristics that can be incorporated into inference to improve the off-nominal behavior of this annotation without actually changing its meaning. these include:

* execute an `@pure` function only if the return type was not perfectly inferred.
* execute `@pure` functions for arguments that subtype `Type` more than for other argument types
* implement execution time step limits

again, this is in contrast to the `@generated` functions and special-case functions that this PR replaces, for which the compiler has no flexibility except to blindly execute at compile time and hope it terminates / works.

but, this PR was intended to only address a couple of particular cases (and provide a slot to store the `pure` attribute). i'm expecting the gf to provide a more general framework and make better use of this information. in light of that, does this need tweaking now to make it compatible with your plans for that?

@stevengj I agree. However, this can only handle completely trivial expressions like `pow(1, 2)`, and not `x = 1; pow(x, 2)`, and it doesn't propagate the information in such a way that it could hoist this out of a loop. I'm not saying you can't mark any function with `Base.@pure`, but this just hasn't implement any of the associated compiler optimizations that make it useful for efficient library functions (outside of the type domain), or any of the safety measures that carnval is describing.